include/swift/SIL/SILType.h

```diff
@@ -586,7 +592,42 @@ class SILType {
 
   /// Returns true if this is the AnyObject SILType;
   bool isAnyObject() const { return getASTType()->isAnyObject(); }
-  
+
+  /// Returns true if this SILType is a move only wrapper type.
+  ///
+  /// Canonical way to check if a SILType is move only. Using is/getAs/castTo
+  /// will look through moveonly-ness.
+  bool isMoveOnly() const { return getRawASTType()->is<SILMoveOnlyType>(); }
+
+  /// Return *this if already move only... otherwise, wrap the current type
+  /// within a move only type wrapper and return that. Idempotent!
+  SILType asMoveOnly() const {
+    if (isMoveOnly())
+      return *this;
+    auto newType = SILMoveOnlyType::get(getRawASTType());
+    return SILType::getPrimitiveType(newType, getCategory());
+  }
+
+  /// Return this SILType, removing moveonly-ness.
+  ///
+  /// Is idempotent.
+  SILType withoutMoveOnly() const {
+    if (!isMoveOnly())
+      return *this;
+    auto moveOnly = getRawASTType()->castTo<SILMoveOnlyType>();
+    return SILType::getPrimitiveType(moveOnly->getInnerType(), getCategory());
+  }
```

Comment:
I don't think this needs to be a public API either. You can't `withoutMoveOnly` a type that is inherently move-only and doesn't need the wrapper to make it so.